File: aeds02/units/u04/lista/Lista.c
#include<stdio.h>
#include<stdlib.h>

#define MAXTAM 6

void inserirInicio(int x);
void inserirFim(int x);
void inserir(int x, int pos);
void mostrar();

int n;
int arr[MAXTAM];

void start() {
	n = 0;
}

void inserirInicio(int x) {
	int i;
	if(n >= MAXTAM) {
		exit(1);
	}
	for(i = n; i > 0; i--) {
		arr[i] = arr[i - 1];
	}
	arr[0] = x;
	n++;
}

void inserirFim(int x) {
	if(n >= MAXTAM) {
		exit(1);
	}
	arr[n] = x;
	n++;
}

void inserir(int x, int pos) {
	int i;
	if(n >= MAXTAM || pos > n || pos < 0) {
		exit(1);
	}
	for(i = n; i > pos; i--) {
		arr[i] = arr[i - 1];
	}
	arr[pos] = x;
	n++;
}

int removerInicio() {
	int i, resp;
	if(n == 0) {
		exit(1);
	}
	resp = arr[0];
	n--;
	for(i = 0; i < n; i++) {
		arr[i] = arr[i + 1];
	}
	return resp;
}

int removerFim() {
	int resp;
	if(n == 0) {
		exit(1);
	}
	return arr[--n];
}

int remover(int pos) {
	int i, resp;
	if(n == 0 || pos > n || pos < 0) {
		exit(1);
	}
	resp = arr[pos];
	n--;
	for(i = pos; i < n; i++) {
		arr[i] = arr[i + 1];
	}
	return resp;
}

void mostrar() {
	int i;
	printf("[ ");
	for(i = 0; i < n; i++) {
		printf("%d ", arr[i]);
	}
	printf("]");
}

int main() {
	int w1, w2, w3;

	start();
	inserirInicio(4);
	inserirFim(6);
	w1 = removerInicio();
	inserirFim(3);
	inserirInicio(7);
	inserir(1, 1);
	w2 = removerFim();
	w3 = remover(2);
	mostrar();
	printf("\nw1 = %d; w2 = %d; w3 = %d;\n", w1, w2, w3);
	return 0;
}

File: aeds02/units/u04/lista/Lista.java
class Lista {
	private int[] array;
	private int n;

	Lista() {
		this(6);
	}

	Lista(int tamanho) {
		this.array = new int[tamanho];
		this.n = 0;
	}

	public void inserirInicio(int el) throws Exception {
		if(n >= this.array.length) 
			throw new Exception("Elementos adicionais no array sao invalidos");
		for(int i = n; i > 0; i--) {
			this.array[i] = this.array[i - 1];
		}
		this.array[0] = el;
		n++;
	}

	public void inserirFim(int el) throws Exception {
		if(n >= this.array.length)
			throw new Exception("Elementos adicionais no array sao invalidos");
		this.array[n] = el;
		n++;
	}

	public void inserir(int el, int pos) throws Exception {
		if(n >= this.array.length || pos < 0 || pos >= this.array.length)
			throw new Exception("Posicao invalida ou elementos adicionais no array");
		for(int i = n; i > pos; i--) {
			this.array[n] = this.array[n - 1];
		}
		this.array[pos] = el;
		n++;	
	}

	public int removerInicio() throws Exception {
		if(n == 0) throw new Exception("Sem elementos presentes no array");
		int resp = this.array[0];
		for(int i = 0; i < n; i++) {
			this.array[i] = this.array[i + 1];
		}
		n--;
		return resp;
	}

	public int removerFim() throws Exception {	
		if(n == 0) throw new Exception("Sem elementos presentes no array");
		return this.array[--n];
	}

	public int remover(int pos) throws Exception {
		if(n == 0 || pos < 0 || pos > n)
			throw new Exception("Sem elementos ou posicao excedente");
		int resp = this.array[pos];
		n--;
		for(int i = pos; i < n; i++) {
			this.array[i] = this.array[i + 1];
		}
		return resp;
	}

	public void mostrar() {
		System.out.print("[ ");
		for(int i = 0; i < n; i++) {
			System.out.print(this.array[i]  + " ");
		}
		System.out.print("]");
	}
}	

class App {
	public static void main(String[] args) {
		Lista lista = new Lista(6);
		int w1 = 0, w2 = 0, w3 = 0;

		try {
			lista.inserirInicio(4);
			lista.inserirFim(5);
			lista.inserir(2, 1);
			w1 = lista.removerInicio();
			w2 = lista.removerFim();
			lista.inserirFim(7);
			lista.inserirInicio(9);
			lista.inserir(12, 2);
			w3 = lista.remover(2);
		} catch(Exception e) { System.out.println(e); }

		lista.mostrar();

		System.out.println("\nw1 = " + w1);
		System.out.println("w2 = " + w2);
		System.out.println("w3 = " + w3);

	}
}
